examples/fluids/qfunctions/stg_shur14.h

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Implementation of the Synthetic Turbulence Generation (STG) algorithm
+/// presented in Shur et al. 2014
+//
+/// SetupSTG_Rand reads in the input files and fills in STGShur14Context. Then
+/// STGShur14_CalcQF is run over quadrature points. Before the program exits,
+/// TearDownSTG is run to free the memory of the allocated arrays.
+
+#ifndef stg_shur14_h
+#define stg_shur14_h
+
+#include <math.h>
+#include <ceed.h>
+#include <stdlib.h>
+
+#ifndef stg_shur14_struct
+#define stg_shur14_struct
+typedef struct STGShur14Context_ *STGShur14Context;
+struct STGShur14Context_ {
+  CeedInt nmodes;              // !< Number of wavemodes
+  CeedScalar *sigma, *d, *phi; // !< Random number set, [nmodes,3], [nmodes,3], [nmodes]
+  CeedScalar *kappa;           // !< Wavemode frequencies, [nmodes]
+  CeedScalar alpha;            // !< Geometric growth rate of kappa
+
+  CeedInt nprof;       // !< Number of profile points in STGInflow.dat
+  CeedScalar *prof_dw; // !< Distance to wall for Inflow Profie, [nprof]
+  CeedScalar *ubar;    // !< Mean velocity, [nprof, 3]
+  CeedScalar *cij;     // !< Cholesky decomposition [nprof, 6]
+  CeedScalar *eps;     // !< Turbulent Disspation [nprof, 6]
+  CeedScalar *lt;      // !< Tubulent Length Scale [nprof, 6]
+};
+#endif
+
+int SetupSTG_Rand(STGShur14Context stg_ctx){
+
+  //TODO will probably want to have the paths to the STGRand.dat and
+  // STGInflow.dat as inputs for this function.
+
+  stg_ctx->alpha = 1.01; // Get from CLI, yaml/toml, etc.
+
+  //TODO Read STGRand.dat
+  stg_ctx->nmodes = 2;
+  stg_ctx->sigma = malloc(sizeof(CeedScalar)*stg_ctx->nmodes*3);
```

Comment:
I'm not understanding why you would be mallocing inside of a QFunction? I don't think that will be GPU friendly.

Reply:
This QFunction code will get called multiple times for the same application of the FE operator in most backends, so I really don't think this is the approach you want to take. This will leak like a sieve.

Reply:
This isn't a QFunction. It's a "normal" function that will be called only once in the program. Though I can see now that it being in the `qfunctions` directory might lead to the confusion. Would `src` would be a better place to move this (and `TearDownSTG`)?

Reply:
Ah, that makes more sense. It would be best to move it. It's best to put only QFunctions in these QFunction headers. It makes JiT easier. 

This still will be a problem on the GPU though. The context data can't be copied to the GPU with a struct that only has pointers and not the data.

Reply:
Yeah, better to put it in `src` -- that's where setup stuff goes. This header will end up being read by the JIT so it should be as slim as possible.

Reply:
That said, we'd rather this be an array inside the `struct` (the last one can have dynamic size, allocated as `malloc(sizeof(struct S) + n*sizeof(double))`) instead of a pointer to other data. The reason being that when a context is copied to the GPU, we'd rather it be plain data rather than needing deep copies.